File: python/patterns-for-coding/fruits_into_baskets.py
"""
Problem Statement #
Given an array of characters where each character represents a fruit tree, you are given two baskets and your goal is to put maximum number of fruits in each basket. The only restriction is that each basket can have only one type of fruit.

You can start with any tree, but once you have started you can’t skip a tree. You will pick one fruit from each tree until you cannot, i.e., you will stop when you have to pick from a third fruit type.

Write a function to return the maximum number of fruits in both the baskets.

Example#1
Input: Fruit=['A', 'B', 'C', 'A', 'C']
Output: 3
Explanation: We can put 2 'C' in one basket and one 'A' in the other from the subarray ['C', 'A', 'C']

Example#2
Input: Fruit=['A', 'B', 'C', 'B', 'B', 'C']
Output: 5
Explanation: We can put 3 'B' in one basket and two 'C' in the other basket. 
This can be done if we start with the second letter: ['B', 'C', 'B', 'B', 'C']
"""
import unittest

def fruits_into_baskets(fruits):
   window_start = 0
   fruits_basket = {}
   max_length = 0

   for window_end in range(len(fruits)):
      right_fruit = fruits[window_end]

      if right_fruit not in fruits_basket:
         fruits_basket[right_fruit] = 0
      fruits_basket[right_fruit] += 1



      while len(fruits_basket) > 2:
         left_fruit = fruits[window_start]

         if left_fruit in fruits_basket:
            fruits_basket[left_fruit] -= 1
         
         if fruits_basket[left_fruit] == 0:
            del fruits_basket[left_fruit]

         window_start += 1

      max_length = max(max_length, window_end - window_start + 1)
   
   return max_length






class Test(unittest.TestCase):
   def test_fruits_into_baskets(self):
      result = fruits_into_baskets(fruits=['A', 'B', 'C', 'A', 'C'])

      print(result)
      assert result == 3
   
   def test_fruits_into_baskets1(self):
      result = fruits_into_baskets(fruits=['A', 'B', 'C', 'B', 'B', 'C'])

      print(result)
      assert result == 5

unittest.main(verbosity=2)

File: js/patterns-for-conding/fruits_into_baskets.js
/** https://www.educative.io/courses/grokking-the-coding-interview/Bn2KLlOR0lQ
 * Problem Statement#
Given an array of characters where each character represents a fruit tree, you are given two baskets, and your goal is to put maximum number of fruits in each basket. The only restriction is that each basket can have only one type of fruit.

You can start with any tree, but you can’t skip a tree once you have started. You will pick one fruit from each tree until you cannot, i.e., you will stop when you have to pick from a third fruit type.

Write a function to return the maximum number of fruits in both baskets.

Example 1:

Input: Fruit=['A', 'B', 'C', 'A', 'C']
Output: 3
Explanation: We can put 2 'C' in one basket and one 'A' in the other from the subarray ['C', 'A', 'C']

Example 2:

Input: Fruit=['A', 'B', 'C', 'B', 'B', 'C']
Output: 5
Explanation: We can put 3 'B' in one basket and two 'C' in the other basket. 
This can be done if we start with the second letter: ['B', 'C', 'B', 'B', 'C']

 */

const fruitsIntoBasket = (fruits) => {
  console.log('in the func')
  let winStart = 0
  let fruitFrequency = {}
  let max = 0

  for (let winEnd = 0; winEnd < fruits.length; winEnd++) {
    let temp = fruits[winEnd]
    
    if (fruitFrequency[temp] === undefined) {
      fruitFrequency[temp] = 1
    } else {
      fruitFrequency[temp] += 1
    }
    console.log('fruitFrequency ', fruitFrequency)
    
    while (Object.keys(fruitFrequency).length > 2) {
      let temp = fruits[winStart]
      console.log('temp while loop', temp)
      
      fruitFrequency[temp] -= 1
  
      if (fruitFrequency[temp] === 0) {
        delete fruitFrequency[temp]
      }
      winStart += 1
      
    }
    max = Math.max(max, winEnd - winStart + 1)
  }
  // console.log(fruits.slice(winStart, winEnd))
  return max
}

let Fruit=['A', 'B', 'C', 'A', 'C']

const res1 = fruitsIntoBasket(Fruit) //Output: 3

console.log('res1 ', res1)

